### PR TITLE
HSEARCH-1970 Rebased again

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://hibernate.atlassian.net/browse/HSEARCH-1970

Also see #896 and #1042 which both are superseded by this one. It's rebased onto my changes related to mass indexing, as some parts were touched in both PRs and I'm expecting the latter to be merged soon (So only the commits for HSEARCH-1970 are the meat of this PR).

@Sanne, I've reworked the things we discussed:

* Added by `IndexManager#getSerializer()` to keep the contract stable. It's not used, though and the known impls return `null`
* Added `SearchIntegrator#getWorkSerializer()` to facilitate retrieval of the serializer without getting penalized for repeated service look ups

To recap on a very high level for those interested: This work changes how `IndexManager` and `BackendQueueProcessor` interact with each other. While it was a circular dependency between impls of the two contracts before, now only BQP "knows" IM but not the other way around. Two concepts are separated more cleanly now:

* How work is propagated and applied: locally, remotely via JMS or remotely via JGroups; Task of BQP impls
* Applying indexing work to the actual indexes: deal with embedded Lucene, or ES, (or Solr); Task of IM

### CODE REVIEW

**Comment by Sanne** [comment]
@gsmet why the "Need wait for other PR" label? I was hoping to merge this soon, but I'll admit I don't have an up to date overview on all other open PRs.

File: engine/src/main/java/org/hibernate/search/indexes/spi/IndexManager.java [hunk 1]
```diff
@@ -121,14 +120,15 @@
...
 	 */
+	@Deprecated
 	LuceneWorkSerializer getSerializer();
 
 	/**
 	 * Flushes any pending write operations and closes underlying resources, for example
 	 * release the index locks if this implementation uses a Lucene IndexWriter.
 	 */
 	void flushAndReleaseResources();
-
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/indexes/spi/IndexManager.java]: FYI this blank line removal caused a confusing rebase conflict

**Comment by gsmet** [comment]
@Sanne because it now incorporates the changes of https://github.com/hibernate/hibernate-search/pull/1094.

**Comment by gunnarmorling** [comment]
Yeah, I opted for basing this one on the bulk indexing work as some code
was touched in both PRs and rebasing/merging would become nightmare-ish
otherwise.

**Comment by Sanne** [comment]
@gunnarmorling sorry, but this needs another rebase.. your fault as you renamed stuff ;-)

**Comment by gunnarmorling** [comment]
Rebased and force-pushed, @Sanne.

**Comment by Sanne** [comment]
thanks @gunnarmorling , merged!

**Comment by gunnarmorling** [comment]
Awesome, very nice to see this one in!
